lib/Platform/Intl/PlatformIntlApple.mm

```diff
@@ -24,6 +24,159 @@
+  // 1. Let candidate be locale
+  std::u16string candidate = locale;
+
+  // 2. Repeat
+  while (true) {
+    // a. If availableLocales contains an element equal to candidate, return
+    // candidate.
+    if (llvh::find(availableLocales, candidate) != availableLocales.end()) {
+      return candidate;
+    }
+
+    // b. Let pos be the character index of the last occurrence of "-" (U+002D)
+    // within candidate.
+    size_t pos = candidate.rfind(u'-');
+
+    // ...If that character does not occur, return undefined.
+    if (pos == std::string::npos) {
+      return llvh::None;
+    }
+
+    // c. If pos ≥ 2 and the character "-" occurs at index pos-2 of candidate,
+    // decrease pos by 2.
+    if (pos >= 2 && candidate[pos - 2] == '-') {
+      pos -= 2;
+    }
+    // d. Let candidate be the substring of candidate from position 0,
+    // inclusive, to position pos, exclusive.
+    candidate.resize(pos);
+  }
+}
+
+// Implementer note: For more information review
+// https://402.ecma-international.org/7.0/#sec-unicode-locale-extension-sequences
+std::u16string toNoUnicodeExtensionsLocale(const std::u16string &locale) {
+  std::vector<std::u16string> subtags;
+  auto s = locale.begin(), e = locale.end();
+  while (true) {
+    auto tagEnd = std::find(s, e, u'-');
+    subtags.emplace_back(s, tagEnd);
+    if (tagEnd == e)
+      break;
+    s = tagEnd + 1;
+  }
+  std::u16string result;
+  size_t size = subtags.size();
+  for (size_t s = 0; s < size;) {
+    if (s > 0) {
+      result.append(u"-");
+    }
+    result.append(subtags[s]);
+    s++;
+    // If next tag is a private marker and there are remaining tags
+    if (subtags[s] == u"u" && s < size - 1)
+      // Skip those tags until you reach end or another singleton subtag
+      while (s < size && subtags[s].size() > 1)
+        s++;
+  }
+  return result;
+}
+// Implementer note: This method corresponds roughly to
+// https://tc39.es/ecma402/#sec-lookupmatcher
+struct LocaleMatch {
+  std::u16string locale;
+  std::u16string extension;
+};
+LocaleMatch lookupMatcher(
+    const std::vector<std::u16string> &requestedLocales,
+    const std::vector<std::u16string> &availableLocales) {
+  // 1. Let result be a new Record.
+  LocaleMatch result;
+  // 2. For each element locale of requestedLocales, do
+  for (const std::u16string &locale : requestedLocales) {
+    // a. Let noExtensionsLocale be the String value that is locale with
+    // any Unicode locale extension sequences removed.
+    std::u16string noExtensionsLocale = toNoUnicodeExtensionsLocale(locale);
+    // b. Let availableLocale be BestAvailableLocale(availableLocales,
+    // noExtensionsLocale).
+    llvh::Optional<std::u16string> availableLocale =
+        bestAvailableLocale(availableLocales, noExtensionsLocale);
+    // c. If availableLocale is not undefined, then
+    if (availableLocale) {
+      // i. Set result.[[locale]] to availableLocale.
+      result.locale = std::move(*availableLocale);
```

Comment:
```suggestion
      result.locale = *availableLocale;
```
`*availableLocale` is an rvalue already. Isn't this cast unnecessary?

Reply:
`*availableLocale` is an lvalue reference, you can see the implementation of `operator*` [here](https://github.com/facebook/hermes/blob/2219f4c5a0c55480088392deab47c000393b0181/external/llvh/include/llvh/ADT/Optional.h#L187). I think you're referring to [this one](https://github.com/facebook/hermes/blob/2219f4c5a0c55480088392deab47c000393b0181/external/llvh/include/llvh/ADT/Optional.h#L196), which returns an rvalue reference if the operator is invoked on an rvalue reference.

Concretely, this allows the below statements to be interchangeable:
```
std::move(*availableLocale)
*std::move(availableLocale)
```
In the first case, we get an lvalue reference from `operator*` on an lvalue and then `std::move` that. In the second case, we get an rvalue reference from `*` because it is invoked on an rvalue (and therefore triggers the rvalue version you're referring to).